dag-cbor/src/encode.rs

```diff
@@ -218,8 +218,11 @@ impl WriteCbor for Cid {
     #[inline]
     fn write_cbor<W: Write>(&self, w: &mut W) -> Result<()> {
         write_tag(w, 42)?;
+        // insert zero byte per https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md#links
         let bytes = self.to_bytes();
-        bytes.as_slice().write_cbor(w)?;
+        write_u64(w, 2, (bytes.len() + 1) as u64)?;
```

Comment:
What about not writing the bytes manually but doing something like:

```rust
        let slice: &[u8] = &[&[0], &bytes[..]].concat();
        slice.write_cbor(w)?;
```

Reply:
Hmm I didn't know of https://doc.rust-lang.org/std/primitive.slice.html#method.concat ... What does it do exactly? It cannot return a slice like in your example, and I would prefer a single additional call to a writer in favor of allocating a new `Vec<_>`. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d6174c92b48b9fa208b3d00741aa60bf shows it returning a Vec.

Reply:
I've tried it locally, all tests pass.

Reply:
That's more like what happens in this case: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5fa868c33ca5f63e310e69d60fad7527

Reply:
> What does it do exactly?

I should answer your question. It concatenates slices.

Reply:
Isn't the additional write call one fewer heap allocation?

Reply:
Also @vmx just a side note: I'm wondering if there could be some construct for this in `rust-cid` i.e. some `Raw` format for this case?

Reply:
> It concatenates slices.

Yes, understood. But as it does it by copying them into a new `Vec`, which requires an allocation, I was thinking it shouldn't be done here. I was initially surprised by the suggestion as I didn't see the `let slice: &[u8] = &expr;`, the ampersand on the right hand side and was confused as to how could a concatenated slice be constructed so (since a slice is a pointer and length).

I guess the point could be made that `Write::write_all` can end up allocating (in case it's a `std::io::Cursor<Vec<u8>>` for example) but at least it will be amortized. When writing to a file this can cause a syscall to be made but then again, `BufWriter` is usually used to alleviate if that becomes a problem.

Reply:
> Also @vmx just a side note: I'm wondering if there could be some construct for this in `rust-cid` i.e. some `Raw` format for this case?

The "raw" format is what `to_bytes()` is doing. You could use multibase to add the `0x0` prefix, but I think this overkill for this simple case.

Reply:
Before getting into tech details. Would you agree that what I propose is kind of cleaner as it shows that we prefix the CID with `0x0` and we use existing methods to encode it (the one to encode slices into CBOR bytes), so it's less duplication of code?

Reply:
Concat will result in an additional heap allocation, I agree with @koivunej, just my 2cents...

Reply:
@vmx Yes your suggestion would definetly be less duplication.

I wager that this code doesn't have much pressure to be changed over time, unless of course the specs change, but still there are tests, so I felt confident selecting the less-allocating route.

Reply:
I generally favour easier to read code and less potential bugs over micro-optimizations in case there are no benchmarks that are based on real-world performance problems. But it's fine for me to keep as it currently is. This was just a suggestion.

Reply:
I agree with the sentiment, however the codec is performance critical as every change to a block requires recursively reencoding with the updated hash...

Reply:
also I believe we do have benchmarks, so if someone would like to run them... :)

Reply:
Thanks for the suggestion. My opinion here is/was also biased because I think allocations are something which the compiler cannot optimize through (as in, it cannot ever remove an allocation) and so we should try to be mindful of those when possible. TBH I haven't really checked for a while if that is accurate.